src/Erl/Data/Map/Internal.purs

```diff
@@ -0,0 +1,190 @@
+  map f m = mapImpl f m
+
+instance foldableMap :: Foldable (Map a) where
+  foldr f z m = foldr f z (values m)
+  foldl f = fold (\z _ -> f z)
+  foldMap f = foldMap (const f)
+
+instance foldableWithIndexMap :: FoldableWithIndex a (Map a) where
+  foldrWithIndex f = fold (\b i a -> f i a b)
+  foldlWithIndex f = fold (\b i a -> f i b a)
+  foldMapWithIndex = foldMap
+
+instance traversableMap :: Traversable (Map a) where
+  traverse f ms = fold (\acc k v -> flip (insert k) <$> acc <*> f v) (pure empty) ms
+  sequence = sequenceDefault
+
+-- | Create a map with one key/value pair
+singleton :: forall a b. a -> b -> Map a b
+singleton a b = insert a b empty
+
+-- | Look up a value for the specified key
+lookup :: forall a b. a -> Map a b -> Maybe b
+lookup = lookupImpl Nothing Just
+
+mapWithKey :: forall k a b. (k -> a -> b) -> Map k a -> Map k b
+mapWithKey f m = mapWithKeyImpl (mkFn2 f) m
+
+-- | Insert the value, delete a value, or update a value for a key in a map
+alter :: forall k v. (Maybe v -> Maybe v) -> k -> Map k v -> Map k v
+alter f k m = lookup k m # f # maybe' (\_ -> delete k m) (\v -> insert k v m)
+
+-- | Update or delete the value for a key in a map
+update :: forall k v. Ord k => (v -> Maybe v) -> k -> Map k v -> Map k v
+update f k m = alter (maybe Nothing f) k m
+
+-- | Fold the keys and values of a map
+fold :: forall a b z. (z -> a -> b -> z) -> z -> Map a b -> z
+fold = foldMImpl ((#))
+
+-- | Fold the keys and values of a map, accumulating values using some
+-- | `Monoid`.
+foldMap :: forall a b m. Monoid m => (a -> b -> m) -> Map a b -> m
+foldMap f = fold (\acc k v -> f k v <> acc) mempty
+
+-- | Fold the keys and values of a map, accumulating values and effects in
+-- | some `Monad`.
+foldM :: forall a b m z. Monad m => (z -> a -> b -> m z) -> z -> Map a b -> m z
+foldM f z = foldMImpl bind f (pure z)
+
+-- | Convert any foldable collection of key/value pairs to a map.
+-- | On key collision, later values take precedence over earlier ones.
+fromFoldable :: forall f k v. Ord k => Foldable f => f (Tuple k v) -> Map k v
+fromFoldable = foldl (\m (Tuple k v) -> insert k v m) empty
+
+-- | Convert any foldable collection of key/value pairs to a map.
+-- | On key collision, the values are configurably combined.
+fromFoldableWith :: forall f k v. Ord k => Foldable f => (v -> v -> v) -> f (Tuple k v) -> Map k v
+fromFoldableWith f = foldl (\m (Tuple k v) -> alter (combine v) k m) empty where
+  combine v (Just v') = Just $ f v v'
+  combine v Nothing = Just v
+
+-- | Convert any indexed foldable collection into a map.
+fromFoldableWithIndex :: forall f k v. Ord k => FoldableWithIndex k f => f v -> Map k v
+fromFoldableWithIndex = foldlWithIndex (\k m v -> insert k v m) empty
+
+-- | Convert a map to an unfoldable structure of key/value pairs where the keys are in ascending order
+toUnfoldable :: forall f k v. Unfoldable f => Map k v -> f (Tuple k v)
+toUnfoldable = L.toUnfoldable <$> toUnfoldMImp Tuple
```

Comment:
related to the FFI traversing the structure multiple times to get into an `Unfoldable`